feat: allow editing for guild role icons

example-git opened the pull request:
## Pull Request Type
<!-- Check the appropriate option -->

- [X] Feature addition
- [ ] Bugfix
- [ ] Documentation update
- [ ] Code refactor
- [ ] Tests improvement
- [ ] CI/CD pipeline enhancement
- [ ] Other: [Replace with a description]

## Description
I have implemented handling into the role.edit method to allow for guild.icon updating, as well as adding the required converter methods into the utils.misc_utils.

## Changes
- added icon and unicode_emoji to the role.edit method. If a unicode emoji is selected it will use that value instead of an icon and vice versa.
- added bytes_to_base64_data and get_mime_type_for_image to utils.misc_utils to handle the bytes imput from the icon input into the proper formatting in the api request.

## Related Issues
<!-- If this PR is related to any open issues, please mention them using "#ISSUENUMBER" -->


## Test Scenarios
<!-- Provide clear instructions on how to test your changes, where applicable - if no tests are required, explain why -->


## Python Compatibility
<!-- Testing 3.11 is not strictly required, but it would be nice if you could confirm that it works. -->
- [X] I've ensured my code works on Python `3.10.x`
- [X] I've ensured my code works on Python `3.11.x`


## Checklist
<!-- If you have not completed all of the following, there is a good chance your PR will not be merged. -->
- [X] I've run the `pre-commit` code linter over all edited files
- [X] I've tested my changes on supported Python versions
- [X] I've added tests for my code, if applicable
- [X] I've updated / added  documentation, where applicable

silasary committed:
Merge pull request #1524 from interactions-py/unstable

5.9.2

LordOfPolls committed:
Merge pull request #1545 from interactions-py/unstable

5.10.0

LordOfPolls committed:
chore: Version bump

Signed-off-by: LordOfPolls <dev@lordofpolls.com>

example-git committed:
Implemented role icon edit

example-git committed:
added role icon editing to the edit method

example-git committed:
Update misc_utils.py

example-git committed:
Merge branch 'unstable' into role-icon-editing

pre-commit-ci[bot] committed:
ci: correct from checks.

example-git committed:
fixed formatting, added return formatting indicator

example-git committed:
Update role.py

i0bs commented:
I don't want to give a review without understanding how this works. Are you able to provide valid client-facing code with your proposal for me?

example-git committed:
Noticed that interactions.client.utils.serializer

already had the functions I added to misc_utils. Undid those changes and tidied up the code in role.py to reflect the already existing converter.  Added an error raise if both icon and unicode emojis are used in one edit (the api requires only one is provided.)

example-git commented:
Here is a snippet of a listener I have that will update a role to match the icon (if one is present) on a role it's tracking. @i0bs 

I also updated the branch because I found that there was already a method in interactions.client.utils.serializer that did exactly what I added to misc_utils. Let me know if the snippet below will suffice.

```py
    async def color_check(self, user):
        roles = user.roles  # remove @everyone
        # highest order of the colour is the one that gets rendered.
        # if the highest is the default colour then the next one with a colour
        # is chosen instead
        role_color = None
        # print(user.top_role.color.value)
        for role in reversed(roles):
            if role.color.value:
                role_color = role.color
                role_name = role.name
        return role_color, role_name
        
    @listen()
    async def on_role_update(self, ctx: RoleUpdate):
        if int(ctx.after.guild.id) != int(self.GuildID):
            return
        user = await ctx.after.guild.fetch_member(self.UserID)
        roles = user.roles
        if int(ctx.after.id) not in roles:
            return
        try:
            self.bot.logger.info(f"Updating Role color to match updated user roles")
            my_role = await ctx.after.guild.fetch_role(self.RoleID)
            role_color, role_name = await self.color_check(user)
            role_icon = None
            try:
                for role in user.roles:
                    role = await ctx.guild.fetch_role(role.id)
                    if role.icon:
                        role_icon = await role.icon.fetch()
                        break
            except:
                role_icon = None
            if not role_icon:
                await my_role.edit(name=role_name, color=role_color)
            else:
                await my_role.edit(name=role_name, color=role_color, icon=role_icon)

i0bs reviewed (changes requested):
Please revert type signatures for `icon` and `unicode_emoji` within the `.edit()` docstring. This is currently inconsistent with the way we already document argument signatures for methods.

Please rewrite both the documentation for `unicode_emoji` to be clearer, simpler and shorter in synopsis. Better choice of verbiage and structure lets us keep things streamlined and straight to the point.

example-git committed:
Cleaned up the docstring and moved the icon serialization

cleaned up the docstring to better align with requirements, moved the icon serialization so the payload's code is a bit easier to follow at a glance.

example-git commented:
> Please revert type signatures for `icon` and `unicode_emoji` within the `.edit()` docstring. This is currently inconsistent with the way we already document argument signatures for methods.
> 
> Please rewrite both the documentation for `unicode_emoji` to be clearer, simpler and shorter in synopsis. Better choice of verbiage and structure lets us keep things streamlined and straight to the point.

How do the changes in the latest commit look

silasary commented on interactions/models/discord/role.py:
```diff
@@ -187,6 +186,8 @@ async def edit(
         color: Color | COLOR_TYPES | None = None,
         hoist: bool | None = None,
         mentionable: bool | None = None,
+        icon: bytes | None = None,
```
`to_image_data` takes UPLOADABLE_TYPE, so you may as well accept it here

silasary reviewed (commented):
Other than an overly narrow type hint, looks good to me

example-git committed:
Added UPLOADABLE_TYPE per comment

LordOfPolls reviewed (approved):
lgtm
